### PR TITLE
Historical and MiddleManager server announcements should not remove parents.

### PR DESCRIPTION
**Pull request by gianm**
Removing parent paths causes watchers of the "announcements" path to get stuck
and stop seeing new updates.

this problem can be reproduced in a totally fresh setup by:

Start a coordinator
Start a historical- it will announce itself and create /druid/announcements
Stop that historical- it will unannounce itself and delete /druid/announcements
The coordinator's pathChildrenCache of /druid/announcements will now never update, even if you start more historicals
There is a similar problem with overlords and middleManagers.

### CODE REVIEW

File: server/src/test/java/io/druid/curator/announcement/AnnouncerTest.java [hunk 1]
```diff
@@ -231,4 +231,29 @@ public void testLeavesBehindTurdlingsThatAlreadyExisted() throws Exception
...
+
+    final byte[] billy = "billy".getBytes();
+    final String testPath = "/somewhere/test2";
+    final String parent = ZKPaths.getPathAndNode(testPath).getPath();
+
+    announcer.start();
+    try {
+      Assert.assertNull(curator.checkExists().forPath(parent));
+
+      announcer.announce(testPath, billy, false);
```
Comment (drcrallen) [inline server/src/test/java/io/druid/curator/announcement/AnnouncerTest.java]: this is not synchronous. Can you have some means of retrying until you are sure curator is caught up?

File: indexing-service/src/main/java/io/druid/indexing/worker/WorkerCuratorCoordinator.java [hunk 2]
```diff
@@ -117,8 +117,6 @@ public void stop() throws Exception
       if (!started) {
         return;
       }
-
-      announcer.unannounce(getAnnouncementsPathForWorker());
```
Comment (drcrallen) [inline indexing-service/src/main/java/io/druid/indexing/worker/WorkerCuratorCoordinator.java]: Why is this removed?

File: server/src/main/java/io/druid/curator/announcement/Announcer.java [hunk 3]
```diff
@@ -62,7 +62,7 @@
   private final CuratorFramework curator;
   private final PathChildrenCacheFactory factory;
 
-  private final List<Pair<String, byte[]>> toAnnounce = Lists.newArrayList();
+  private final List<Announceable> toAnnounce = Lists.newArrayList();
   private final List<Pair<String, byte[]>> toUpdate = Lists.newArrayList();
```
Comment (drcrallen) [inline server/src/main/java/io/druid/curator/announcement/Announcer.java]: Can this also use `Announceable`?

File: server/src/main/java/io/druid/curator/announcement/Announcer.java [hunk 4]
```diff
@@ -367,14 +376,30 @@ private void startCache(PathChildrenCache cache)
     }
   }
 
-  private void createPath(String parentPath)
+  private void createPath(String parentPath, boolean removeParentsIfCreated)
   {
     try {
       curator.create().creatingParentsIfNeeded().forPath(parentPath);
```
Comment (drcrallen) [inline server/src/main/java/io/druid/curator/announcement/Announcer.java]: Is there a way to log if the parent was created by this call or not?

**Comment by drcrallen** [comment]
@gianm can you expand a little more on when this error is encountered?

File: indexing-service/src/main/java/io/druid/indexing/worker/WorkerCuratorCoordinator.java (hunk 2 above)
Comment (gianm) [inline indexing-service/src/main/java/io/druid/indexing/worker/WorkerCuratorCoordinator.java]: Unannounce is called on stop

File: server/src/main/java/io/druid/curator/announcement/Announcer.java (hunk 3 above)
Comment (gianm) [inline server/src/main/java/io/druid/curator/announcement/Announcer.java]: It could although the removeParentsIfCreated part isn't relevant. I can just set it to false though.

File: server/src/main/java/io/druid/curator/announcement/Announcer.java (hunk 4 above)
Comment (gianm) [inline server/src/main/java/io/druid/curator/announcement/Announcer.java]: could add a log.debug, does that work?

Comment (drcrallen) [inline server/src/main/java/io/druid/curator/announcement/Announcer.java]: Yes, that would be good

File: server/src/test/java/io/druid/curator/announcement/AnnouncerTest.java (hunk 1 above)
Comment (gianm) [inline server/src/test/java/io/druid/curator/announcement/AnnouncerTest.java]: sure, I had just copied an older test so I will fix that one too. It looks like 'announce' does background ops but 'unannounce' does foreground ops- so we need to wait for 'announce' but not 'unannounce'.

**Comment by gianm** [comment]
@drcrallen pushed changes

**Comment by gianm** [comment]
@drcrallen this problem can be reproduced in a totally fresh setup by:

1. Start a coordinator
2. Start a historical- it will announce itself and create /druid/announcements
3. Stop that historical- it will unannounce itself and delete /druid/announcements
4. The coordinator's pathChildrenCache of /druid/announcements will now never update, even if you start more historicals

There is a similar problem with overlords and middleManagers.

**Comment by drcrallen** [comment]
@gianm can you update the master comment with this information?

**Comment by drcrallen** [comment]
Is this severe enough to shove into 0.8.2?

**Comment by gianm** [comment]
@drcrallen moved to the top comment

I think it's not a regression, so probably not urgent from that POV, but probably worth including if we are gonna do another rc anyway.

**Comment by drcrallen** [comment]
:+1:

**Comment by drcrallen** [comment]
@gianm can you get an 0.8.2 backport PR going?

**Comment by gianm** [comment]
@drcrallen backport pr is #1870
